File: zmq-reqrep/mtserver.c
//  Multithreaded REQ-REP server

#include "zhelpers.h"
#include <pthread.h>
#include <unistd.h>

#define DATA_UNIT_SIZE 1024

static char *buffer = NULL;

static void *
worker_routine (void *context) {
    //  Socket to talk to dispatcher
    void *receiver = zmq_socket (context, ZMQ_REP);
    zmq_connect (receiver, "inproc://workers");

    while (1) {
        char *string = s_recv (receiver);
        // printf ("Received request: [%s]\n", string);
        free (string);
        //  Send reply back to client
        s_send (receiver, buffer);
    }
    zmq_close (receiver);
    return NULL;
}

int main (int argc, char **argv)
{
    size_t bufsize = DATA_UNIT_SIZE;
    int multiple = 1;

    if (argc >= 2) {
        multiple = atoi(argv[1]);
        bufsize *= multiple;
    }

    buffer = (char *) malloc(bufsize + 1);
    memset(buffer, 0, bufsize + 1);

    int i = 0;
    for (; i < bufsize; i++)
        buffer[i] = 'z';

    printf("Set buffer size to %ld\n", bufsize);
    // printf("%ld %s\n", strlen(buffer), buffer);

    void *context = zmq_ctx_new ();

    //  Socket to talk to clients
    void *clients = zmq_socket (context, ZMQ_ROUTER);
    zmq_bind (clients, "tcp://*:5555");

    //  Socket to talk to workers
    void *workers = zmq_socket (context, ZMQ_DEALER);
    zmq_bind (workers, "inproc://workers");

    //  Launch pool of worker threads
    int thread_nbr;
    for (thread_nbr = 0; thread_nbr < 5; thread_nbr++) {
        pthread_t worker;
        pthread_create (&worker, NULL, worker_routine, context);
    }
    //  Connect work threads to client threads via a queue proxy
    zmq_proxy (clients, workers, NULL);

    //  We never get here, but clean up anyhow
    zmq_close (clients);
    zmq_close (workers);
    zmq_ctx_destroy (context);
    return 0;
}

File: zmq-reqrep/mtserver.py
"""

   Multithreaded Reqrep server

   Author: Guillaume Aubert (gaubert) <guillaume(dot)aubert(at)gmail(dot)com>

   Author: Junda Feng

"""
import time
import threading
import zmq

WORKER_NUM = 5
DEFAULT_SIZE = 1024

BUFFERS = {}


def setup_buffers():
    """Prepare responses"""

    import sys, string
    buffer_size = DEFAULT_SIZE
    if len(sys.argv) >= 2:
        buffer_size = int(sys.argv[1])

    print("Set buffer size to {}".format(buffer_size))

    for letter in string.ascii_lowercase:
        BUFFERS[letter] = letter * buffer_size


def worker_routine(worker_url, context=None):
    """Worker routine"""

    context = context or zmq.Context.instance()
    # Socket to talk to dispatcher
    socket = context.socket(zmq.REP)
    socket.connect(worker_url)

    while True:
        try:
            request = socket.recv_string()
            #send reply back to client
            socket.send_string(BUFFERS[request])
        except zmq.ContextTerminated:
            break

def main():
    """Server routine"""

    url_worker = "inproc://workers"
    url_client = "tcp://*:5555"

    setup_buffers()

    # Prepare our context and sockets
    context = zmq.Context.instance()

    # Socket to talk to clients
    clients = context.socket(zmq.ROUTER)
    clients.bind(url_client)

    # Socket to talk to workers
    workers = context.socket(zmq.DEALER)
    workers.bind(url_worker)

    # Launch pool of worker threads
    for i in range(WORKER_NUM):
        thread = threading.Thread(target=worker_routine, args=(url_worker,))
        thread.start()

    try:
        zmq.device(zmq.QUEUE, clients, workers)
    except KeyboardInterrupt:
        pass

    # We never get here but clean up anyhow
    clients.close()
    workers.close()
    context.term()


if __name__ == "__main__":
    main()
